wagtail/admin/rich_text/converters/editor_html.py

```diff
@@ -143,7 +143,9 @@ def html_rewriter(self):
             elif isinstance(rule, LinkTypeRule):
                 link_rules[rule.link_type] = rule.handler.expand_db_attributes
 
-        return MultiRuleRewriter([LinkRewriter(link_rules), EmbedRewriter(embed_rules)])
+        return MultiRuleRewriter(
+            [LinkRewriter(rules=link_rules), EmbedRewriter(rules=embed_rules)]
```

Comment:
@chosak should these not be `bulk_rules` a la https://github.com/wagtail/wagtail/pull/5875/files#diff-28e86f41edadc0bbf445852a6284888ab58ce7b2947725fccf60fbd189085698R30? or am I forgetting something?

Reply:
It's confusing!

The richtext rewriters LinkRewriter and EmbedRewriter currently (in main, before this PR) have a signature like

```py
LinkRewriter(rules, reference_extractors=None)
```

where `rules` is a mapping of `{linktype: expand_db_attributes}`, for example, given a linktype `page`, call method `PageLinkHandler.expand_db_attributes`. Each method takes a dict of attributes `{"linktype": "page", "id": 1234}` and returns the object being referenced (e.g. the page with ID 1234).

What this PR does is add another bulk approach to doing these lookups, such that the signature now looks like

```py
LinkRewriter(rules=None, bulk_rules=None, reference_extractors=None)
```

If bulk rules are provided, they will be used instead; for example this PR defines a `PageLinkHandler.expand_db_attributes_many` that allows for bulk lookup of page objects given multiple attribute dicts.

This PR reimplements all of the _DB -> frontend HTML_ rewriters so that they have bulk implementations. It's possible that devs have implemented other non-bulk custom rewriters, especially since this interface has been documented.

The _rich text editor -> DB_ rewrite path uses different rewrite rules (named similarly, unfortunately) that that this PR has not rewritten to work in bulk. I could try to migrate these to work in bulk, but I'm not sure the implementation works similarly enough for that to make sense (note for example that [these are `WhitelistRule` instances](https://github.com/wagtail/wagtail/blob/c11db31a787ab68df79bb051838dc51633a8b2ee/wagtail/admin/rich_text/converters/editor_html.py#L124-L127) that do a lot of cleanup that doesn't involve fetching things from the database).

In other words, the two different rewrite paths seemingly use the same framework (MultiRuleRewriter, LinkRewriter, EmbedRewriter) to accomplish different tasks.

Reply:
You're right. Despite my double checking, I muddled this up again. Long day!

Thank you for taking the time to spell it out clearly. May I suggest a developer note just above `MultiRuleRewriter()` here to hint and warn future us of the confusing naming (e.g. "The rich text editor -> DB rewrite rules are different than the DB -> front end ones. The latter accept `bulk_rules` kwarg for bulk lookups"

Reply:
I think aside from this, and the decision on https://github.com/wagtail/wagtail/pull/5875#discussion_r1493739982 on whether we 

a. do nothing and convert rules to bulk rules
b. add a deprecation notice for custom rewriters
c. just drop support for non-bulk rules for obscure rewriters since you handle it for `LinkHandler` and `EmbedHandler`s

Reply:
Hi @zerolab, just checking back in on this. How can I help move this PR along?

Reply:
Assigned to @gasman for this last bit.